### DISCUSSION TITLE
[GS] Two Modest Proposals about GS and NewGRF

### DISCUSSION THREAD

**Discussion by andythenorth** [OpenTTD/OpenTTD#8460] [Features]
**1. Permit GS - NewGRF communication via registers**

_GS - NewGRF co-operation about data_

* towns already have registers that grfs can read and write to
* GS gains ability to read / write to town registers
* grfs gain a grf-global set of registers 
* GS gains ability to read / write to grf registers

Who gets to write, and who gets to read-only, I will leave to people better at this than me.

**2. Permit packaging a GS with a grf**

_Reducing friction for players to use co-operative GS + NewGRF_

Assumes something like Capability API pre-exists https://wiki.openttd.org/en/Development/Design%20Drafts/Scripts/Capability-based%20API

* when players activate a grf that requires a GS, that GS is activated for them automatically
* or vice versa if a GS requires a grf
* GS is somehow scoped to the grf

How to package 2 types of co-dependent content, I will leave to people better at this than me.

_See also_ 

"A modest proposal" is usually intended as satire 😄 https://en.wikipedia.org/wiki/A_Modest_Proposal

I don't intend these proposals as satire.  They just seem quite...hard 😸

**Comment by Andrew350** [comment]
I like it, especially point 2.

**My specific usecase:** Currently NewGRFs can change town growth behavior by changing cargo acceptance of houses, but only GS has access to changing the town information window. This leaves no way to effectively communicate to players your new town growth rules through NewGRF-only; it requires a 'companion' GS to add that information. _(This is assuming the rule about keeping NewGRF and GS domains separate still holds, such that NewGRF won't gain this functionality)_ 

GS can also be useful for adding background or other gameplay hints through the storybook for more complicated setups, along with a host of other game-enhancing changes for a specific NewGRF. Although this is all possible now, it requires that players are aware of and activate your GS and NewGRF together, which can't be relied upon.

Andy's point 1 would also potentially open up lots of possibilities, though what exactly would likely depend on implementation, and I don't have a specific usecase in mind at the moment.

**Comment by andythenorth** [comment]
See also GS Identities https://gist.github.com/TrueBrain/fe3c80e7dc9313b7ac7941540e5dee22

**Comment by JGRennison** [comment]
I think 2 is a good idea.
The constraints of multiplayer would limit how closely the two could really be coupled.
At present GRFs are run on all clients, but GSs are only run on the server.
GRFs are required to be deterministic and run in bounded time but GSs are not, so the two would be a bit at arms length from each other.

That said, Squirrel is much easier to read, write, debug and extend than NFO/GRF.
(If NFO was proposed today, there would no be no bargepole in the world long enough to bat it away).
Allowing more of the complex functionality to be specified in readable Squirrel instead of having to use NFO would likely be a positive move.

----

On 1, I think that the principle of communication between mods (GSs, GRFs, etc.) is a good idea, but I think that the mechanism of shared *writable* registers is a bit problematic.

Adding persistent registers scoped to a GRF as a whole seems fairly easy and inoffensive, but it seems likely that they'd also end up as another temporary scratchpad for computation/switches.

Allowing GSs or other GRFs to write into the registers owned by another GRF is (in my opinion) not a good idea, and should probably not be allowed.

GRFs are already quite opaque and difficult to debug. Unexpected interactions between unrelated GRFs are also already a problem.
Allowing other mods to poke at the memory (registers, etc.) of GRFs would probably worsen this.
Also nobody wants to receive a bug report that their mod is broken because some other mod scribbled over their registers, even if they chose to allow that.

Giving GSs their own "GRFID" and their own registers which only they can exclusively write to would probably be better, (except for pre-emption issues, see below).

GSs have different behaviour to GRFs:
GRFs are currently run serially, with deterministic ordering. As a result GRFs reading each others registers is pretty much OK.
GSs are (effectively) run by pre-emptive task switching.
A GS which does:
```
read register 0
read register 1
```
could be interrupted between the two lines for an arbitrary length of time.
The game could advance and GRFs could update the values of both of the registers between those two lines, such that the GS is operating off inconsistent data. This problem is worse if the GS is allowed to write to externally readable registers as well.
Non-deterministic timing-related bugs are likely to be frustrating for mod developers and users.

Mod developers may want to communicate things larger than a single register, and may want to do things like send a request with data and receive a response in another register.
e.g. https://wiki.openttd.org/en/Development/Script/Script%20communication%20protocol except with registers instead of signs.
Getting this right in a pre-emptible context like this is surprisingly difficult. This is especially the case if there are more than two entities involved.

A few other thoughts:
Registers do not currently have names. Registers are pretty much meaningless without documentation of what each one is. It would be nice if there was a standard place or format for this, or even if it could be surfaced in debug UIs.
The NewGRF debug window has code for town persistent registers, but I cannot work out how to drive it to make it actually show them. Does this even work at all and if not how are mod developers supposed to debug this?
Using registers for more things makes it more important that it is possible to find out which mods are using them, and when/how.
At the moment there's really no way to find out if a mod is reading and dependant on some other mod's registers.

**Comment by andythenorth** [comment]
Thanks JGR for such a detailed reply 👍 

Pre-emption issues seem difficult.  

I am wary of adding confusion by using the wrong terminology, but it seems to me that authors using GS<->grf communication need to design it to be asynchronous, more akin to message queues or similar.  

Examples use cases that might work:
* GS designates a set of adjacent towns as 'coal mining town', writes that to a town register, grf then uses that as a signal when constructing industries 
  * done once at game start
* GS writes price factor per cargo into town register, which grf custom cargo profit callback uses for 'demand-based' payment rates  
  * for example use combination of town cargo monitor and indications from town grf industries to modify 'demand'
  * bonus: modify cdist demand function using same approach
  * runs 'monthly' (has to be designed on basis that it may not complete once a month)
* grf industry writes to town register 'please try and build an industry of type X in this town' GS then tries to construct industry of that type in town
  * runs 'soon' but exact timeframe for completion shouldn't matter

Whereas e.g. delegating grf industry production callback handling to GS would be architecturally just not possible.  This would be fine.

Re: named registers.  Same issue applies when authoring grf.  They can probably be named as constants in nml, but I skipped that and named them in the FIRS compile https://github.com/andythenorth/firs/blob/master/src/industry.py#L1577

These are for the industry registers, which are not exposed externally, and vary per industry type, so it's just an FYI, not anything we could build on. 

I do have ideas for FIRS -> house grf communication, which will use the town registers, and _will_ need a convention for which register holds what. 

I wondered about reserving some registers to provide something like a schema (mapping conventional labels to register numbers).  Probably too complex right now though.

I wondered about exposing a schema in action 14, but that would be grf-global, and rely on authors not changing the schema per-town. I am way out of my depth on that 😄 

I haven't tried accessing the town register via grf debug window, but I will need to for the FIRS -> house grf idea.

_Aside_ I've only suggested registers for this as they're an existing structure and we understand them.  There could be other ways.
